File: product-microservice/api/product_api.go
package api

import (
	"product/mapper"
	"product/middleware"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type ProductAPI struct {
	IProductService service.IProductService
}

func NewProductAPI(productService service.IProductService) ProductAPI {
	return ProductAPI{IProductService: productService}
}

func (productApi *ProductAPI) GetProductById(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	product, err := productApi.IProductService.GetProductById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToProductDTO(product))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (productApi *ProductAPI) AddProductToMainPage(c *gin.Context) {
	productId, err := strconv.Atoi(c.Param("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := productApi.IProductService.AddProductToMainPage(productId)

	if msg == "" {
		c.JSON(http.StatusOK, "Product added to main page.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (productApi *ProductAPI) RemoveProductFromMainPage(c *gin.Context) {
	productId, err := strconv.Atoi(c.Param("productId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := productApi.IProductService.RemoveProductFromMainPage(productId)

	if msg == "" {
		c.JSON(http.StatusOK, "Product removed from main page.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (productApi *ProductAPI) DeleteProduct(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	msg := productApi.IProductService.DeleteProduct(id)

	if msg == "" {
		c.JSON(http.StatusOK, "Product deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (productApi *ProductAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	products, err := productApi.IProductService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToProductDTOs(products))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (productApi *ProductAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := productApi.IProductService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (productApi *ProductAPI) GetMainPageProducts(c *gin.Context) {
	products := productApi.IProductService.GetMainPageProducts()
	c.JSON(http.StatusOK, mapper.ToProductDTOs(products))
}

func (productApi *ProductAPI) GetPopularProducts(c *gin.Context) {
	products := productApi.IProductService.GetPopularProducts()
	c.JSON(http.StatusOK, mapper.ToProductDTOs(products))
}

func (productApi *ProductAPI) GetRecommendedProducts(c *gin.Context) {
	userData := middleware.GetUserData(c)
	products := productApi.IProductService.GetRecommendedProducts(userData.Id)
	c.JSON(http.StatusOK, mapper.ToProductDTOs(products))
}

File: api-gateway/api/product_microservice/product_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import product_api_routes


def get_product_by_id(id):
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.GET_PRODUCT_BY_ID + "/{}".format(id))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.SEARCH_BY_NAME +
                         "?page={}&pageSize={}&name={}".format(page, page_size, name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(err)
        resp.status_code = 404
        return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_main_page_products():
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.GET_MAIN_PAGE_PRODUCTS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_popular_products():
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.GET_POPULAR_PRODUCTS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def add_product_to_main_page(productId):
    headers = request.headers
    try:
        r = requests.put(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.ADD_PRODUCT_TO_MAIN_PAGE + "/{}".format(productId), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(err)
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def remove_product_from_main_page(productId):
    headers = request.headers
    try:
        r = requests.put(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.REMOVE_PRODUCT_FROM_MAIN_PAGE + "/{}".format(productId), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete_product(id):
    headers = request.headers
    try:
        r = requests.delete(product_api_routes.BASE + product_api_routes.API +
                            product_api_routes.DELETE_PRODUCT + "/{}".format(id), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_USER])
def get_recommended_products():
    headers = request.headers
    try:
        r = requests.get(product_api_routes.BASE + product_api_routes.API +
                         product_api_routes.GET_RECOMMENDED_PRODUCTS, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp
